File: frontend/src/api/nutrition.ts
// Nutrition API. Attributes go in the POST body (not the query string) so
// personal data stays out of URLs.

import { api } from './client'
import type { ActivityLevel, NutritionGoal, NutritionTargets } from './types'

export interface NutritionRequest {
  sex?: 'male' | 'female' | 'other'
  age: number
  heightCm: number
  weightKg: number
  activity: ActivityLevel
  goal: NutritionGoal
}

interface NutritionTargetsPayload {
  bmr: number
  tdee: number
  calories: number
  protein_g: number
  carbs_g: number
  fat_g: number
  bmi: number
  bmi_category: string
  goal: NutritionGoal
  warning: string | null
}

export async function calculateNutrition(req: NutritionRequest): Promise<NutritionTargets> {
  const p = await api.post<NutritionTargetsPayload>('/nutrition/calculate', {
    sex: req.sex,
    age: req.age,
    height_cm: req.heightCm,
    weight_kg: req.weightKg,
    activity: req.activity,
    goal: req.goal,
  })
  return {
    bmr: p.bmr,
    tdee: p.tdee,
    calories: p.calories,
    proteinG: p.protein_g,
    carbsG: p.carbs_g,
    fatG: p.fat_g,
    bmi: p.bmi,
    bmiCategory: p.bmi_category,
    goal: p.goal,
    warning: p.warning,
  }
}

File: backend/app/api/v1/schemas/nutrition.py
"""Pydantic schemas for the nutrition endpoints."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.domain.value_objects.enums import ActivityLevel, NutritionGoal


class NutritionRequest(BaseModel):
    """Body attributes for the calorie/macro estimate. Sent in the POST body
    (not the query string) so personal data stays out of URLs."""

    sex: Literal["male", "female", "other"] | None = None
    age: int = Field(ge=14, le=100)
    height_cm: float = Field(ge=120, le=230)
    weight_kg: float = Field(ge=30, le=300)
    activity: ActivityLevel
    goal: NutritionGoal


class NutritionTargetsRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    bmr: int
    tdee: int
    calories: int
    protein_g: int
    carbs_g: int
    fat_g: int
    bmi: float
    bmi_category: str
    goal: NutritionGoal
    warning: str | None = None
